plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs

```diff
@@ -0,0 +1,364 @@
+import Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types.Lens as J
+import Retrie.ExactPrint (Annotated)
+import RnSplice
+import TcRnMonad
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId =
+    (defaultPluginDescriptor plId)
+        { pluginCommands = commands
+        , pluginCodeActionProvider = Just codeAction
+        }
+
+--
+expandInplaceId, expandCommentedId :: CommandId
+expandInplaceId = "expandTHSpliceInplace"
+expandCommentedId = "expandTHSpliceCommented"
+
+inplaceCmdName :: T.Text
+inplaceCmdName = "expand TemplateHaskell Splice (in-place)"
+
+commentedCmdName :: T.Text
+commentedCmdName = "expand TemplateHaskell Splice (comented-out)"
+
+commands :: [PluginCommand]
+commands =
+    [ PluginCommand expandInplaceId inplaceCmdName $ expandTHSplice Inplace
+    -- , PluginCommand expandCommentedId commentedCmdName $ expandTHSplice Commented
+    ]
+
+expandTHSplice ::
+    -- | Inplace?
+    ExpandStyle ->
+    CommandFunction ExpandSpliceParams
+expandTHSplice eStyle lsp ideState params@ExpandSpliceParams {..} =
+    fmap (fromMaybe defaultResult) $
+        runMaybeT $ do
+            fp <- MaybeT $ pure $ uriToNormalizedFilePath $ toNormalizedUri uri
+            TcModuleResult {tmrParsed = pm, ..} <-
+                MaybeT $
+                    runAction "expandTHSplice.TypeCheck" ideState $
+                        use TypeCheck fp
+            ps <-
+                MaybeT $
+                    useAnnotatedSource
+                        "expandTHSplice.AnnotedSource"
+                        ideState
+                        fp
+            hscEnvEq <-
+                lift $
+                    runAction "expandTHSplice.ghcSessionDeps" ideState $
+                        use_ GhcSessionDeps fp
+            let hscEnv0 = hscEnvWithImportPaths hscEnvEq
+                modSum = pm_mod_summary pm
+            hscEnv <- lift $
+                evalGhcEnv hscEnv0 $ do
+                    env <- getSession
+                    df <- liftIO $ setupDynFlagsForGHCiLike env $ ms_hspp_opts modSum
+
+                    let impPaths = fromMaybe (importPaths df) (envImportPaths hscEnvEq)
+
+                    -- Set the modified flags in the session
+                    _lp <- setSessionDynFlags df {importPaths = impPaths}
+
+                    -- copy the package state to the interactive DynFlags
+                    idflags <- getInteractiveDynFlags
+                    setInteractiveDynFlags $
+                        idflags
+                            { pkgState = pkgState df
+                            , pkgDatabase = pkgDatabase df
+                            , packageFlags = packageFlags df
+                            , useColor = Never
+                            , canUseColor = False
+                            }
+                    env' <- getSession
+                    -- setTargets [thisModuleTarget]
+                    resl <- load LoadAllTargets
```

Comment:
Why do this? `load` will effectively double IDE memory usage.

Reply:
At first, I couldn't make out how to make symbols in splice into Renamer's scope and I struggled to resolve such situations just by stealing some codes from Eval Plugin.
At last, I found that these and passing `TcGblEnv` to `init*Tc` resolves things, but, yes, I must definitely try to remove some/all fragments of these HscEnv setting hackeries. I'll try, Thanks!

Reply:
It seems we don't need `evalGhcWith` hacks isn't needed entirely, so I deleted that part. Thanks again!